### PR TITLE
Add S3 Copy URL

### PR DESCRIPTION
**Pull request by abrooksv**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
Add Copy URL to S3 object context menu. Generating presigned is out of scope for this PR

## Related Issue(s)
#2082 

## Screenshots (if appropriate)
Screen Shot 2020-09-24 at 11 52 26 AM

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .changes/next-release/feature-8ae7facb-e1c5-4f25-9955-8c36b3cf59c4.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "type" : "feature",
+  "description" : "Added the ability to copy the URL to an S3 object"
```
Comment (rli) [inline .changes/next-release/feature-8ae7facb-e1c5-4f25-9955-8c36b3cf59c4.json:3]: Should we standardize our descriptions? I can never remember if we're trying to make everything past or present tense

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3VirtualBucketTest.kt [hunk 2]
```diff
@@ -197,4 +200,16 @@ class S3VirtualBucketTest {
...
+    @Test
+    fun getUrl() {
+        val s3Client = mockClientManager.create<S3Client>()
+
+        s3Client.stub {
+            on { utilities() } doReturn S3Utilities.builder().region(Region.US_WEST_2).build()
+        }
+
+        val sut = S3VirtualBucket(Bucket.builder().name("TestBucket").build(), s3Client)
+        assertThat(sut.generateUrl("prefix/key")).isEqualTo(URL("https://s3.us-west-2.amazonaws.com/TestBucket/prefix/key"))
```
Comment (rli) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3VirtualBucketTest.kt:213]: what about the virtual-hosted path style?

Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3VirtualBucketTest.kt:213]: We have to enable that on our S3 client and we don't do that at all today.

File: .changes/next-release/feature-8ae7facb-e1c5-4f25-9955-8c36b3cf59c4.json (hunk 1 above)
Comment (abrooksv) [reply .changes/next-release/feature-8ae7facb-e1c5-4f25-9955-8c36b3cf59c4.json:3]: Ya "Add" does sound better

File: core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt [hunk 3]
```diff
@@ -15,7 +15,8 @@ class DelegateSdkConsumers : Answer<Any> {
     override fun answer(invocation: InvocationOnMock): Any? {
         val method = invocation.method
         return if (method.isDefault &&
-            method?.parameters?.getOrNull(0)?.type?.kotlin?.isSubclassOf(SdkRequest::class) != true
+            method?.parameters?.getOrNull(0)?.type?.kotlin?.isSubclassOf(SdkRequest::class) != true &&
+            method.name !in setOf("utilities", "waiter")
```
[the remark this answers is not in the document]
Comment (abrooksv) [reply core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt:19]: No, cause it shames me

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3VirtualBucketTest.kt [hunk 4]
```diff
@@ -197,4 +199,14 @@ class S3VirtualBucketTest {
         assertThat(request.prefix()).isEqualTo("prefix/")
         assertThat(request.delimiter()).isEqualTo("/")
     }
+
+    @Test
+    fun getUrl() {
```
[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3VirtualBucketTest.kt:205]: Added a notification

File: core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (abrooksv) [reply core/tst/software/aws/toolkits/core/utils/DelegateSdkConsumers.kt:19]: Will add

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyUrlAction.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import java.awt.datatransfer.StringSelection
+
+class CopyUrlAction(private val project: Project, treeTable: S3TreeTable) : SingleS3ObjectAction(treeTable, message("s3.copy.url")) {
+    // Only enable it if we have some selection. We hide the root node so it means we have no selection if that is the node passed in
+    override fun enabled(node: S3TreeNode): Boolean = node != treeTable.rootNode
+
+    override fun performAction(node: S3TreeNode) {
+        try {
+            CopyPasteManager.getInstance().setContents(StringSelection(treeTable.bucket.generateUrl(node.key).toString()))
+            S3Telemetry.copyUrl(project, presigned = false)
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyUrlAction.kt:22]: We should add `result` to the telemetry

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/CopyUrlAction.kt:22]: https://github.com/aws/aws-toolkit-common/pull/68

**Review by hunterwerlla — APPROVED** [review]

**Pull request merged**